diff --git a/mlir/lib/Dialect/MemRef/Utils/MemRefUtils.cpp b/mlir/lib/Dialect/MemRef/Utils/MemRefUtils.cpp
index 4879c5cc98ea..a4eb2b2e2c32 100644
--- a/mlir/lib/Dialect/MemRef/Utils/MemRefUtils.cpp
+++ b/mlir/lib/Dialect/MemRef/Utils/MemRefUtils.cpp
@@ -160,16 +160,18 @@ static bool resultIsNotRead(Operation *op, std::vector<Operation *> &uses) {
 }
 
 void eraseDeadAllocAndStores(RewriterBase &rewriter, Operation *parentOp) {
-  std::vector<Operation *> opToErase;
+  std::SetVector<Operation *> opToErase;
   parentOp->walk([&](memref::AllocOp op) {
     std::vector<Operation *> candidates;
     if (resultIsNotRead(op, candidates)) {
-      opToErase.insert(opToErase.end(), candidates.begin(), candidates.end());
-      opToErase.push_back(op.getOperation());
+      for(auto candidate: candidates)
+        opToErase.insert(candidate);
+      opToErase.insert(op.getOperation());
     }
   });
   for (Operation *op : opToErase) {
-    rewriter.eraseOp(op);
+    if (op->use_empty())
+      rewriter.eraseOp(op);
   }
 }
 